### PR TITLE
Documentation and tests for IteratorDecorator.

### PR DESCRIPTION
**Pull request by erikvanoosten**
If this is approved, I will add similar docs and tests for the other intersperse implementations.

### CODE REVIEW

**Comment by erikvanoosten** [comment]
I just realized that the correct term for things in a collection is `element` and not `item`. I will update this PR.

**Comment by erikvanoosten** [comment]
I completed docs and unit tests for `IteratorDecorator`.

I did not document `lazyFoldLeft` and `lazyFoldRight` since I found not find words to describe their exotic behavior:

```
    def op(acc: Int, e: => Int): Int = if (acc >= 5) acc + 1 else acc + e
    Assert.assertEquals(11, Iterator(4, 5, 20, 20).lazyFoldLeft(0)(op))
```
I'll work some more on this project in a new PR.
Comments welcome!

**Comment by julienrf** [comment]
> I had expected we could use `$consumesIterator` and `$mayNotTerminateInf` in the scaladocs...

I suggest that you copy their definitions from the standard library. https://github.com/scala/scala/blob/05d028e0eee1eaf9032e652395b94a124c6ad9ed/src/library/scala/collection/Iterator.scala#L50

File: src/main/scala/scala/collection/decorators/IteratorDecorator.scala [hunk 1]
```diff
@@ -5,6 +5,20 @@ import scala.annotation.tailrec
 
 class IteratorDecorator[A](val `this`: Iterator[A]) extends AnyVal {
 
+  /**
+    * Inserts a separator value between each element.
+    *
+    * {{{
+    *   Iterator(1, 2, 3).intersperse(0) === Iterator(1, 0, 2, 0, 3)
```
Comment (julienrf) [inline src/main/scala/scala/collection/decorators/IteratorDecorator.scala]: I think that if we give code examples, then this should be code examples that can be copy-pasted. Here the method `===` does not exist, and even `==` wouldn’t work.

Maybe we can just add a note at the beginning explaining that this is “pseudo-code” and that `===` means “is equivalent as if we had written”?

Comment (erikvanoosten) [reply src/main/scala/scala/collection/decorators/IteratorDecorator.scala]: We could write `assert( ... == ... )`, that would be executable.

File: src/main/scala/scala/collection/decorators/IteratorDecorator.scala [hunk 2]
```diff
@@ -5,6 +5,20 @@ import scala.annotation.tailrec
...
+    *   Iterator('a', 'b', 'c').intersperse(',') === Iterator('a', ',', 'b', ',', 'c')
+    *   Iterator('a').intersperse(',') === Iterator('a')
+    *   Iterator().intersperse(',') === Iterator()
+    * }}}
+    *
+    * @param sep the separator value.
+    * @return    The resulting iterator contains all elements from the source iterator, separated by the `sep` value.
+    * @note      Reuse: $consumesIterator
```
Comment (julienrf) [inline src/main/scala/scala/collection/decorators/IteratorDecorator.scala]: I think here we should use `$consumeAndProducesIterator` instead

File: src/main/scala/scala/collection/decorators/IteratorDecorator.scala (hunk 1 above)
Comment (erikvanoosten) [reply src/main/scala/scala/collection/decorators/IteratorDecorator.scala]: Or just `.. == ..`. That would executable also :)

Comment (julienrf) [reply src/main/scala/scala/collection/decorators/IteratorDecorator.scala]: Yes, but `==` is reference equality on `Iterator`, not value equality, so, executing that code would return `false`.

File: src/main/scala/scala/collection/decorators/IteratorDecorator.scala (hunk 2 above)
Comment (erikvanoosten) [reply src/main/scala/scala/collection/decorators/IteratorDecorator.scala]: Yeah, already fixed that :)

File: src/main/scala/scala/collection/decorators/IteratorDecorator.scala (hunk 1 above)
Comment (erikvanoosten) [reply src/main/scala/scala/collection/decorators/IteratorDecorator.scala]: Or: `Iterator(1, 2, 3).intersperse(0) // equals Iterator(1, 0, 2, 0, 3)`

Comment (erikvanoosten) [reply src/main/scala/scala/collection/decorators/IteratorDecorator.scala]: I'll take your suggestion.

Comment (erikvanoosten) [reply src/main/scala/scala/collection/decorators/IteratorDecorator.scala]: Fixed.

File: src/main/scala/scala/collection/decorators/IteratorDecorator.scala [hunk 3]
```diff
@@ -3,8 +3,39 @@ package decorators
 
 import scala.annotation.tailrec
 
+/** Enriches [[Iterator]] with additional methods.
```
Comment (julienrf) [inline src/main/scala/scala/collection/decorators/IteratorDecorator.scala]: It should be just `Iterator`

**Comment by julienrf** [comment]
Thanks a lot! Could you please squash the comits into a single one?

File: src/main/scala/scala/collection/decorators/IteratorDecorator.scala (hunk 3 above)
Comment (erikvanoosten) [reply src/main/scala/scala/collection/decorators/IteratorDecorator.scala]: oh.

**Comment by erikvanoosten** [comment]
Isn't squashing an option when you merge in github? If not, I can find out how to do that.

**Comment by julienrf** [comment]
Right, there is an option for that.

**Pull request merged**